### PR TITLE
Use alias symbol for optional parameter declaration serialization

### LINKED ISSUE #42165: Declaration emit redundantly inlines optional union parameters
# Bug Report

### 🔎 Search Terms

inlining duplication declaration emit optional union parameters

### 🕗 Version & Regression Information

- This is not a regression.
- This is a stylistic issue in emitted DTS.  The code is correct.
- This is the behavior in every version I tried (3.3 ... nightly)

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?#code/C4TwDgpgBArlC8UCMUA+UBMaoGZsBZsBWAbigFgAoKiADzAHsAnYKAYwYDsBnVhhKAG8qUUVAC2AQ04ATScGYgAFJKYBzAFywAlFoBuDAJYyhqtQF8ANKID0NqABEAKgGUoTCADMIHzmwjcsJyGXFBKagwMMtoiYgxgwCGckgA2KuoA-FowulAGxqbqVrb2zm6GnCkVAUFJYQBGktFU5lRAA)

### 💻 Code

```ts
type u = 1 | 2 | 3 | 4 | 5; 

export const o = {
    mandatory(arg: u): void {arg},  // DTS references union (good)
    optional(arg?: u): void {arg},  // DTS inlines union (bad)
}
```

### 🙁 Actual behavior

DTS emit expands the definition of the union for the optional parameter.

```ts
declare type u = 1 | 2 | 3 | 4 | 5;
export declare const o: {
    mandatory(arg: u): void;
    optional(arg?: 1 | 2 | 3 | 4 | 5 | undefined): void;
};
```

### 🙂 Expected behavior

DTS emit should treat it the same as a mandatory parameter and reference it by name.

```ts
declare type u = 1 | 2 | 3 | 4 | 5;
export declare const o: {
    mandatory(arg: u): void;
    optional(arg?: u): void;
};
```

### Related Issues

This is a very specific instance of [the issue of declarations inlining types](https://github.com/microsoft/TypeScript/issues/37151).  I'm raising it because it looks like a pure win to simplify emitted DTS files.  Fixing this will make other userland tricks to simplify types more effective.

If fixing this is contentious, or requires a large overhaul, please feel free to immediately close.  I will not be offend because I realize there is a long tail of issues in this space and do not wish to create noise.

### PR DESCRIPTION
**Pull request by andrewbranch**
Fixes #42165

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -5823,6 +5823,20 @@ namespace ts {
                 return !(getObjectFlags(type) & ObjectFlags.Reference) || !isTypeReferenceNode(existing) || length(existing.typeArguments) >= getMinTypeArgumentCount((type as TypeReference).target.typeParameters);
             }
 
+            function serializeParameterTypeForDeclaration(context: NodeBuilderContext, type: Type, symbol: Symbol, enclosingDeclaration: Node | undefined, includePrivateSymbol?: (s: Symbol) => void, bundled?: boolean) {
+                if (type !== errorType && symbol.valueDeclaration?.kind === SyntaxKind.Parameter && (symbol.valueDeclaration as ParameterDeclaration).questionToken) {
```
Comment (weswigham) [inline src/compiler/checker.ts:5827]: Do we have to look at the value declaration here? Or is there a `CheckFlag` on the symbol which indicates optionality? Moreover, do we already have an `isOptionalParameterSymbol` helper? I can't remember offhand.

**Comment by DanielRosenwasser** [comment]
#42149 seems potentially related here.

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -5823,6 +5823,20 @@ namespace ts {
...
             }
 
+            function serializeParameterTypeForDeclaration(context: NodeBuilderContext, type: Type, symbol: Symbol, enclosingDeclaration: Node | undefined, includePrivateSymbol?: (s: Symbol) => void, bundled?: boolean) {
+                if (type !== errorType && symbol.valueDeclaration?.kind === SyntaxKind.Parameter && (symbol.valueDeclaration as ParameterDeclaration).questionToken) {
+                    const nonOptionalType = getTypeWithFacts(type, TypeFacts.NEUndefined);
+                    if (nonOptionalType !== type && nonOptionalType.aliasSymbol && nonOptionalType.flags & TypeFlags.Union) {
+                        context.approximateLength += 12; // " | undefined"
+                        return factory.createUnionTypeNode([
+                            typeToTypeNodeHelper(nonOptionalType, context),
+                            factory.createKeywordTypeNode(SyntaxKind.UndefinedKeyword)
```
Comment (robpalme) [inline src/compiler/checker.ts:5833]: Do we need to union with `undefined`?  I thought the nullable `?` already implies that.

Comment (andrewbranch) [reply src/compiler/checker.ts:5833]: Declaration emit (or any kind of type printback) always expands optional parameters/properties to their union type annotation, so this is just to match the convention. In some tangent on some other PR I floated the idea of reversing that decision for printback in errors and quick info to make it more succinct, but Anders wasn't keen on it.

At any rate, I'm pretty sure this PR is redundant with #42149. Now that it’s merged, have you tried your test case with it?

Comment (DanielRosenwasser) [reply src/compiler/checker.ts:5833]: Hm - I have always kind of hated the extra `| undefined` unless we've lost the context where it came from.

Comment (andrewbranch) [reply src/compiler/checker.ts:5833]: I think it’s something we should revisit sometime.

**Comment by andrewbranch** [comment]
Superseded by #42149

**Pull request closed without merging**